### PR TITLE
[16501] PKI exchange finished 

### PR DESCRIPTION
**Pull request by Mario-DL**
## Contributor Checklist
- [ ] Commit messages follow the project guidelines. 
- [ ] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [ ] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [ ] Changes are ABI compatible. 
- [ ] Changes are API compatible. 
- [ ] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: test/unittest/rtps/security/CMakeLists.txt [hunk 1]
```diff
@@ -77,6 +77,7 @@ target_include_directories(SecurityAuthentication PRIVATE
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/StatelessReader
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/StatefulReader
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/ReaderHistory
+    ${PROJECT_SOURCE_DIR}/test/mock/rtps/PDP
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/PDPSimple
```
Comment (MiguelCompany) [inline test/unittest/rtps/security/CMakeLists.txt:81]: Perhaps we can remove the PDPSimple mock?

File: src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.hpp [hunk 2]
```diff
@@ -36,22 +36,27 @@ class PDP;
  * Class PDPSecurityInitiatorListener, implementation for the secure discovery server handshake initiator.
  * @ingroup DISCOVERY_MODULE
  */
-class PDPSecurityInitiatorListener: public ReaderListener
+class PDPSecurityInitiatorListener : public ReaderListener
 {
 
+    using ResponseCallback = std::function<void (const ParticipantProxyData& participant_data)>;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.hpp:42]: I think `SecurityInitiatedCallback` is a better name for this

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 3]
```diff
@@ -234,14 +236,133 @@ void PDPServer::update_builtin_locators()
...
+                    remote_readers.emplace_back(participant_data.m_guid.guidPrefix, c_EntityId_SPDPReader);
+
+                    for (auto& locator : participant_data.metatraffic_locators.unicast)
+                    {
+                        locators.push_back(locator);
+                    }
+
+                    send_announcement(change, remote_readers, locators, false);
+
+                }
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:315]:
```suggestion
                    send_announcement(change, remote_readers, locators, false);
                }
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 4]
```diff
@@ -234,14 +236,133 @@ void PDPServer::update_builtin_locators()
...
+
+    endpoints.stateless_listener.reset(new PDPSecurityInitiatorListener(this,
+            [this](const ParticipantProxyData& participant_data)
+            {
+                auto endpoints = static_cast<fastdds::rtps::DiscoveryServerPDPEndpoints*>(builtin_endpoints_.get());
+                std::lock_guard<fastrtps::RecursiveTimedMutex> wlock(endpoints->writer.writer_->getMutex());
+
+                CacheChange_t* change = discovery_db().cache_change_own_participant();
+
+                if (change != nullptr)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:301]:
```suggestion
                CacheChange_t* change = discovery_db().cache_change_own_participant();
                if (change != nullptr)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/unittest/rtps/security/CMakeLists.txt (hunk 1 above)
Comment (Mario-DL) [reply test/unittest/rtps/security/CMakeLists.txt:81]: I am not sure, the SecurityTests rely on  `#include "SecurityTests.hpp"` in which PDPSimple is needed and used in some tests. If needed, we could refactor the tests to use pdp()

Comment (MiguelCompany) [reply test/unittest/rtps/security/CMakeLists.txt:81]: Didn't realize. No need to change them now then. If required, we'll do that on a separate PR.

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Review by MiguelCompany — APPROVED** [review]
LGTM. Let's check that it builds and we'll merge.

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Review by MiguelCompany — APPROVED** [review]

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**